[ML][Maps] Adds link to maps in charts section of Anomaly Explorer

alvarezmelissa87 opened the pull request:
## Summary

Related meta issue: https://github.com/elastic/kibana/issues/123492

This PR adds a link in the anomaly explorer embedded maps to the Maps plugin for displaying the anomalies for that job id.

<img width="779" alt="image" src="https://user-images.githubusercontent.com/6446462/160502867-acb27f7d-5a22-4b9b-847f-df797013dc3c.png">

Links to:

<img width="1191" alt="image" src="https://user-images.githubusercontent.com/6446462/160502890-60c4c894-4e4b-4d8e-9960-d005c12b7b45.png">



### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

nreese commented on x-pack/plugins/maps/public/index.ts:
```diff
@@ -17,7 +17,8 @@ export const plugin: PluginInitializer<MapsPluginSetup, MapsPluginStart> = (
   return new MapsPlugin(initContext);
 };
 
-export { MAP_SAVED_OBJECT_TYPE } from '../common/constants';
```
Instead of exporting APP_ID in maps/public/index, an ML just import from `maps/common`. You will have to add `APP_ID` to `maps/common/index.ts`

alvarezmelissa87 replied:
Good catch - moved to export from /common in 3bfae94c05b9d5ccd1dae877e83cf7a374e59b3e

peteharverson commented on x-pack/plugins/ml/public/application/explorer/explorer_charts/explorer_charts_container.js:
```diff
@@ -79,6 +89,47 @@ function ExplorerChartContainer({
+    }
+
+    const locator = share.url.locators.get(MAPS_APP_LOCATOR);
+    const location = await locator.getLocation({
```
Is it possible to set a query in the target Maps page, to filter the data for the partitioning field(s) in the map clicked on in the Anomaly Explorer? Rather than showing all anomalies for the job, I usually want to visualize location data just for the selected partition (`by` field).

![image](https://user-images.githubusercontent.com/7405507/160601026-94bb19b6-74ae-467e-bde5-411031244413.png)

alvarezmelissa87 replied:
Added query for partition fields in 56bdb872f33bbd964a747f2b7ec923aa44e3ceca

peteharverson commented on x-pack/plugins/ml/public/application/explorer/explorer_charts/explorer_charts_container.js:
```diff
@@ -191,6 +265,23 @@ function ExplorerChartContainer({
               </EuiToolTip>
             )}
+            {chartType === CHART_TYPE.GEO_MAP && mapsLink ? (
+              <EuiToolTip position="top" content={openInMapsPluginMessage}>
```
Not needed for this PR, but would be great to add a `View in Maps` item to the Actions menu in the anomalies table, similar to the `View series` link for time series anomalies which is available for the anomaly charts and the anomalies table.

peteharverson commented on x-pack/plugins/ml/public/application/explorer/explorer_charts/explorer_charts_container.js:
```diff
@@ -79,6 +89,47 @@ function ExplorerChartContainer({
+          }),
+          style: {
+            type: 'VECTOR',
+            properties: {
```
What do you think about using the grey color for the fill / line color for the typical layer, giving the opened map a similar appearance to the one in the Anomaly Explorer?

alvarezmelissa87 replied:
Good call - updated to use grey color for typical layer in 3bfae94c05b9d5ccd1dae877e83cf7a374e59b3e

peteharverson commented on x-pack/plugins/ml/public/maps/anomaly_source.tsx:
```diff
@@ -147,8 +149,23 @@ export class AnomalySource implements IVectorSource {
+
+    if (meta?.areResultsTrimmed) {
+      return {
+        tooltipContent: i18n.translate('xpack.ml.maps.resultsTrimmedMsg', {
```
Not related to this code, but is it possible to only add the Influencers section of the point tooltip if they are not already shown as a partition field. For example here the only influencer I have configured is `vehicle.id` so the empty Influencers section is redundant:

![image](https://user-images.githubusercontent.com/7405507/160603711-ca29c7f8-589d-40f2-a368-de5eb99a2633.png)

nreese reviewed (approved):
maps changes LGTM, changes to anomaly_source.tsx areResultsTrimmed LGTM
code review.

qn895 commented:
Tested latest changes and LGTM but might be worth testing it some more with where partition fields contain special characters

peteharverson reviewed (approved):
Tested latest changes and LGTM

alvarezmelissa87 commented:
> Tested latest changes and LGTM but might be worth testing it some more with where partition fields contain special characters

Added tests for the query creation and added use of `escapeKuery` from `@kbn/es-query` service to ensure special characters are escaped in fd5c131b87593144e31308b14e0dd52681d98837

kibanamachine commented:
Friendly reminder: Looks like this PR hasn’t been backported yet.
To create backports run `node scripts/backport --pr 128697` or prevent reminders by adding the `backport:skip` label.
